applications/nrf_desktop/src/modules/dvfs.c

```diff
@@ -0,0 +1,186 @@
+ */
+enum dvfs_frequency_setting dvfs_freq_table[DVFS_STATE_COUNT] = {
+	DVFS_FREQ_LOW,		/* DVFS_STATE_DEFAULT */
+	DVFS_FREQ_MEDLOW,	/* DVFS_STATE_LLPM */
+	DVFS_FREQ_HIGH		/* DVFS_STATE_USB */
+	};
+
+/* Binary mask tracking which states are requested */
+static uint8_t dfvs_requests_state;
+static enum dvfs_frequency_setting current_freq = DVFS_FREQ_HIGH;
+/* DVFS inits after a while, so we need to wait for it to be ready */
+static struct k_work_delayable DVFS_init_timeout;
+
+static const char *get_dvfs_frequency_setting_name(enum dvfs_frequency_setting setting)
+{
+	switch (setting) {
+	case DVFS_FREQ_HIGH: return "DVFS_FREQ_HIGH";
+	case DVFS_FREQ_MEDLOW: return "DVFS_FREQ_MEDLOW";
+	case DVFS_FREQ_LOW: return "DVFS_FREQ_LOW";
+	default: return "Unknown";
+	}
+}
+
+static void set_dvfs_freq(enum dvfs_frequency_setting target_freq)
+{
+	size_t ret = dvfs_service_handler_change_freq_setting(target_freq);
+
+	if (ret == -EAGAIN) {
+		LOG_INF("DVFS not initialized, try again.");
+	} else if (ret == -EBUSY) {
+		LOG_INF("DVFS frequency change in progress");
+	} else if (ret != 0) {
+		LOG_ERR("DVFS freq change returned with error: %d", ret);
+	} else {
+		LOG_INF("Requesting %s frequency",
+			get_dvfs_frequency_setting_name(target_freq));
+		current_freq = target_freq;
+	}
+}
+
+static enum dvfs_frequency_setting check_required_frequency(void)
+{
+	enum dvfs_frequency_setting highest_requested_freq = DVFS_FREQ_LOW;
+
+	for (int i = 0; i < DVFS_STATE_COUNT; i++) {
+		if (dfvs_requests_state & BIT(i)) {
+			/* We have inversion in the freq DVFS_FREQ_LOW has the highest value */
+			if (highest_requested_freq > dvfs_freq_table[i]) {
+				highest_requested_freq = dvfs_freq_table[i];
+				if (highest_requested_freq == DVFS_FREQ_HIGH) {
+					break;
+				}
+			}
+		}
+	}
+
+	return highest_requested_freq;
+}
+
+static void update_dfvs_requests_state(enum dvfs_states_list state, bool turn_on)
+{
+	if (turn_on) {
+		dfvs_requests_state |= BIT(state);
+	} else {
+		dfvs_requests_state &= ~BIT(state);
+	}
+}
+
+static void handle_dfvs_request(enum dvfs_states_list state, bool turn_on)
+{
+	update_dfvs_requests_state(state, turn_on);
+	enum dvfs_frequency_setting required_freq = check_required_frequency();
+
+	if (required_freq != current_freq) {
+		set_dvfs_freq(required_freq);
+	}
+}
+
+static void check_if_llpm_state(const struct ble_peer_conn_params_event *event)
+{
+	if (!event->updated) {
+		/* Ignore the connection parameters update request. */
+		return;
+	}
+
+	__ASSERT_NO_MSG(event->interval_min == event->interval_max);
+
+	if (event->interval_min & REG_CONN_INTERVAL_LLPM_MASK) {
+		handle_dfvs_request(DVFS_STATE_LLPM, true);
+		LOG_DBG("LLPM_STATE_ACTIVE");
+	} else {
+		handle_dfvs_request(DVFS_STATE_LLPM, false);
+		LOG_DBG("LLPM_STATE_NOT_ACTIVE");
+	}
+}
+
+static void DVFS_init_timeout_handler(struct k_work *work)
+{
+	LOG_DBG("DVFS should be initialized by now.");
+
+	handle_dfvs_request(DVFS_STATE_DEFAULT, true);
```

Comment:
I think it may be more reasonable to rely on app initialization chain here instead (wait for `settings_loader`, `ble_bond`, `ble_adv`/`ble_scan`). After app is fully initialized you can clear the `APP_INITIALIZING` bit

Reply:
disused offline. Will base on retrying mechanism.